src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1470,25 +1477,36 @@ internal void Reset()
             WriteBuffer.Clear();
             _pendingRfqPrependedMessages = 0;
 
-            // Must rollback transaction before sending DISCARD ALL
+            // Must rollback transaction before sending DISCARD commands
             if (InTransaction)
                 Rollback();
 
-            if (SupportsDiscard)
-                PrependInternalMessage(PregeneratedMessage.DiscardAll);
+            // Deallocate non persisted prepared statements
+            if (SupportsDeallocate)
+            {
+                var stmtNames = PreparedStatements.ClearNonPersistedPreparedStatements();
+
+                for (int i = 0; i < stmtNames.Count; i++)
+                {
+                    var query = String.Format("DEALLOCATE \"{0}\";", stmtNames[i]);
+                    PrependInternalMessage(QueryMessage.Populate(query));
```

Comment:
Note that this line fails if there are many deallocations, i.e. if there are enough to fill the 8192 byte send buffer.